chain/chain/src/chain.rs

```diff
@@ -536,22 +533,110 @@ impl Chain {
+    //             \-[F]---[G]
+    //
+    // 1. Let's define clearing height = 102. It this case fork A-F-G is protected from deletion
+    //    because of G which is on height 103. Nothing will be deleted.
+    // 2. Let's define clearing height = 103. It this case Fork Clearing will be executed for A
+    //    to delete blocks G and F, then Fork Clearing will be executed for B to delete block E.
+    //    Then Canonical Chain Clearing will delete blocks A and B as unlocked.
+    //    Block C is the only block of height 103 remains on the Canonical Chain (invariant).
+    //
+    pub fn clear_data(&mut self, trie: Arc<Trie>) -> Result<(), Error> {
         let mut chain_store_update = self.store.store_update();
         let head = chain_store_update.head()?;
+        let tail = chain_store_update.tail()?;
+        // TODO processing super long epochs comes in next updates, possibly in separate PR
         let height_diff = NUM_EPOCHS_TO_KEEP_STORE_DATA * self.epoch_length;
-        if head.height >= height_diff {
-            let last_height = head.height - height_diff;
-            for height in last_height.saturating_sub(HEIGHTS_TO_CLEAR)..last_height {
-                match chain_store_update.clear_old_data_on_height(height) {
+        // TODO build_refcount_map optimization comes in next updates, possibly in separate PR
+        chain_store_update.build_refcount_map()?;
+        if head.height > height_diff {
+            // Forks Cleaning
+            for height in tail..(head.height - height_diff) {
+                if height == 0 {
+                    continue;
+                }
+                match chain_store_update.clear_forks_data(trie.clone(), height) {
                     Ok(_) => {}
                     Err(err) => {
-                        error!(target: "client", "Error clearing old data on height {:?}, {:?}", height, err);
+                        error!(target: "client", "Error at forks clearing on height {:?}, {:?}", height, err);
                     }
                 }
             }
+
+            // Canonical Chain Clearing
+            for height in tail + 1..(head.height - height_diff) {
+                let blocks_current_height: Vec<CryptoHash> =
+                    match chain_store_update.get_all_block_hashes_by_height(height) {
+                        Ok(blocks_current_height) => {
+                            blocks_current_height.values().flatten().cloned().collect()
+                        }
+                        _ => continue,
+                    };
+                assert_ne!(blocks_current_height.len(), 0);
+                if blocks_current_height.len() > 1 {
+                    // Fork is here
+                    break;
+                }
+                let block_hash = blocks_current_height[0];
+                let prev_hash = chain_store_update.get_block_header(&block_hash)?.prev_hash;
+                assert_ne!(chain_store_update.get_block_refcount(&prev_hash), 0);
```

Comment:
same here

Reply:
This assert means something very wrong is happening. I'm not sure about our policy of exiting but I'd suggest to panic here.

Reply:
`debug_assert` is also fine

Reply:
Ok, I've changed to `debug_assert`, reopen if necessary.